GitHub Actions: Add macOS cases.

junaruga opened the pull request:
* Migrate macOS cases in Travis to GitHub Actions.
* Increase the expected time in the test to avoid the following random failures
  on macOS.

  ```
  1) Mysql2::Client#query threaded queries should be supported
     Failure/Error: expect(stop - start).to be_within(0.1).of(sleep_time)
       expected 0.632204999999999 to be within 0.1 of 0.5
     # ./spec/mysql2/client_spec.rb:711:in `block (3 levels) in <top (required)>'
  ```

  ```
  3) Mysql2::Client#query should run signal handlers while waiting for a response
     Failure/Error: expect(mark.fetch(:QUERY_END) - mark.fetch(:QUERY_START)).to be_within(0.1).of(query_time)
       expected 1.1042130000000157 to be within 0.1 of 1.0
     # ./spec/mysql2/client_spec.rb:632:in `block (3 levels) in <top (required)>'
  ```

## Notes

* I renamed `.github/workflows/ubuntu.yml` to `build.yml` again, after adding the macOS cases to the file. I considered the name `ubuntu_macos.yml`. But thinking about we want to keep a same name for a long time, and considering the name (`name: Build`) is used as a build badge name when we will add it later. The name `Build` makes sense to me.

* In Travis macOS case, the brew package `mariadb-connector-c` was installed. So I always installed the `mariadb-connector-c` in GitHub Actions macOS case too. However when we do not actually install it on the `db: 'mysql@5.6'` case of the macOS, `rake compile` can be compiled, and there are some test failures. In `db: 'mariadb@10.4'` case, `rake compile` shows the following library missing error when `mariadb-connector-c` is not installed.
 
  ```
  checking for -lmysqlclient... no
  mysql client is missing. You may need to 'brew install mysql' or 'port install mysql', and try again.
  ```

  We might be able to install `mariadb-connector-c` conditionally on macOS case like this.

  ```
  brew install "${DB}"
  if [[ -n ${DB-} && x$DB =~ ^xmariadb ]]; then
    brew install mariadb-connector-c
  fi
  ```

junaruga commented on .travis_setup.sh:
```diff
@@ -2,23 +2,24 @@
 
 # Install the default used DB if DB is not set.
-if [[ -n ${GITHUB_ACTION-} && -z ${DB-} ]]; then
+if [[ -n ${GITHUB_ACTIONS-} && -z ${DB-} ]]; then
```
* I renamed the environment variable `GITHUB_ACTION` to `GITHUB_ACTIONS`. I think the `GITHUB_ACTIONS` showing always `true` on GitHub Actions environment is a right environment variable, and what I wanted to use. See https://docs.github.com/en/actions/reference/environment-variables .

junaruga commented on .travis_setup.sh:
```diff
@@ -2,23 +2,24 @@
 # where ALTER USER is not available.
 # https://stackoverflow.com/questions/56052177/
-CHANGED_PWD_BY_RECREATE=false
+CHANGED_PASSWORD_BY_RECREATE=false
```
* I renamed `CHANGED_PWD` to `CHANGED_PASSWORD`, because `PWD` makes me imagine `pwd` (print working directory) command. Sorry it was my typo.

junaruga commented on .travis_setup.sh:
```diff
@@ -105,25 +116,20 @@ else
-  fi
-
-  # IF NOT EXISTS is mariadb-10+ only - https://mariadb.com/kb/en/mariadb/comment-syntax/
-  mysql -u root -e 'CREATE DATABASE /*M!50701 IF NOT EXISTS */ test'
```
The new macOS `db: 'mysql@5.6'` case showed the error when running `mysql -u root -e 'CREATE DATABASE /*M!50701 IF NOT EXISTS */ test'`, because it was recognized as `mysql -u root -e 'CREATE DATABASE test'` and there was `test` database already. As every used database supports `IF NOT EXISTS` in CIs, I think we can remove the `/*M!50701 ... */`.

junaruga commented on .github/workflows/build.yml:
```diff
@@ -38,6 +38,12 @@ jobs:
+          # https://github.com/brianmario/mysql2/issues/965
+          - {os: macos-latest, ruby: 2.4, db: 'mariadb@10.4', allow-failure: true}
+          # Allow failure due to Mysql2::Error: Unknown system variable 'session_track_system_variables'.
+          - {os: macos-latest, ruby: 2.4, db: 'mysql@5.6', allow-failure: true}
```
While there was only one case mysql 5.6 in macOS, I added one more diverse case mariadb 10.4 that looks the latest version on the macOS environment, seeing `brew search mariadb`. It is to test the macOS case and the testing logic on GitHub Actions with diverse cases. It seems that macOS is slower than Ubuntu in GitHub Actions. But I think having the 2 macOS cases is still okay.

sodabrew replied:
I think MySQL 5.6 is pretty old at this point. MySQL 8.0 is the way to go.

junaruga replied:
@sodabrew OK I will change it to MySQL 8.0. By the way, I noticed that the brew package `mysql` without `@X.Y` is MYSQL 8.0, and the brew package `mariadb` is 10.5. These are the latest version on the macOS environment. I will change MariaDB to MariaDB 10.5 too.

```
mysql # => This is 8.0
mysql@5.6
mysql@5.7
mariadb => This is 10.5
mariadb@10.1
mariadb@10.2
mariadb@10.3
mariadb@10.4
```

junaruga commented:
The CIs are okay for both GitHub Actions and Travis CI. I noticed the "Checks" tab showing the result of GitHub Actions. It looks convenient.

junaruga commented:
I added one more commit for the review. Now [MariaDB 10.5 (10.5.9)](https://github.com/brianmario/mysql2/pull/1170/checks?check_run_id=1974205505#step:8:68) and [MySQL 8.0 (8.0.23)](https://github.com/brianmario/mysql2/pull/1170/checks?check_run_id=1974205524#step:8:68) are set on the macOS cases. I removed the `mariadb-connector-c` brew package of installing, because it was conflicted with the `mariadb` brew package. And both latest MariaDB and MySQL works without the `mariadb-connector-c`.

* On MariaDB 10.5 (10.5.9), the tests fail with #965, #1152.
* On MySQL 8.0 (8.0.23), the tests fail with #1165.

junaruga commented:
I did split the previous 2 commits to 4 commits, splitting the previous 1st commit about migrating macOS and renaming variables. I think it's better to see the commit about migrating macOS. The CIs looks okay. ([Travis log](https://travis-ci.org/github/junaruga/mysql2/builds/760570858) on my forked repository).

sodabrew commented:
Looks good!
